File: gcsg.py
import numpy as np

def gcsg(fun, x0, T, h=1.0):
    """
    Generalized Centered Simplex Gradient (GCSG) approximation.

    Parameters:
        fun: callable, function from R^n to R, accepts a 1D numpy array
        x0: numpy array, point in R^n (1D)
        T: numpy array, n x m matrix of direction vectors (columns)
        h: float, step size (default=1.0)

    Returns:
        Approximate gradient at x0 as a numpy array of length n.
    """
    x0 = np.array(x0, dtype=float).ravel()
    T = np.array(T, dtype=float)
    n, m = T.shape

    if len(x0) != n:
        raise ValueError("T and x0 must have same number of rows")

    T_h = h * T
    delta_f = np.zeros(m)

    for i in range(m):
        f_forward  = fun(*(x0 + T_h[:, i]))
        f_backward = fun(*(x0 - T_h[:, i]))
        delta_f[i] = (f_forward - f_backward) / 2.0

    grad_approx = np.linalg.pinv(T_h).T @ delta_f
    return grad_approx

File: gcsg.m
%Author: Gabriel Jarry-Bolduc
%Date Created: April, 2020.
%Last Update: May 12, 2025.
%
%This function returns the generalized centered simplex gradient.
%
%Input:fun: function handle from R^n to R
%      x0: point of interest x0. Column vector in R^n
%       T: set of directions in matrix form  n \times m to add to x0
%       h: parameter to shrink T. Positive number.
%
%Output: CSGradValue: gcsg value at x0
%        info: info.AbsError: Absolute error between the true gradient and
%                             the gcsg at x0          
%              info.RelError: Relative error
%              info.gradient: True gradient at x0
%
function [CSGradValue,info] = gcsg(fun,x0,T,h)
%% Set h=1 if it is not specified
if (nargin<3)
   error('gcsg must have at least three inputs');
elseif (nargin==3)
  h=1;
end
%% Compute the gcsg
Sgval1=gsg(fun,x0,T,h);
Sgval2=gsg(fun,x0,-T,h);
CSGradValue=0.5*(Sgval1+Sgval2);
%% Calculate Absolute Error and relative error
if nargout==2
    n=length(x0);
    x=sym('x',[n,1]);
    Fun=fun(x);
    Gradient=gradient(Fun,x);
    for iter=1:n
        CellArrayVar{1,iter}=x(iter);
        CellArray{1,iter}=x0(iter); %  Create a cell since gradfun does not accept a vector as an input
    end
    Gradient_x0=double(subs(Gradient, CellArrayVar, CellArray));
    info.gradient=Gradient_x0';
    info.AbsError=double((norm(CSGradValue-Gradient_x0)));
    Norm=norm(Gradient_x0);
    if Norm>1e-16
        info.RelError=double(norm(CSGradValue-Gradient_x0)/norm(Gradient_x0));
    end
end
end
